File: frontend/src/components/plata/banco.ts
// ─── El libro del banco, del lado del cliente ────────────────────────────────
//
// Los tipos de acá son la forma EXACTA que devuelve `routers/banco.py` (que a su
// vez la relee de `services/banco.py`). No hay un tipo "de pantalla" distinto del
// tipo "de la API" a propósito: cada vez que la UI se inventa una forma propia
// aparece un lugar donde traducirla, y ese lugar es donde después se cuela un
// saldo calculado dos veces con dos reglas distintas. El saldo del banco tiene
// UNA sola matemática y vive en el backend.

/** Un riel por donde entra y sale la plata (Occidente, Bold). */
export interface CuentaBanco {
  id: number
  nombre: string
  activa: boolean
  nota: string | null
}

/** Un movimiento TECLEADO. El monto va siempre positivo: el signo lo pone `tipo`. */
export interface MovimientoBanco {
  id: number
  fecha: string                     // YYYY-MM-DD
  cuenta_id: number
  cuenta: string                    // nombre ya resuelto por el backend
  tipo: 'entrada' | 'salida'
  monto: number                     // SIEMPRE positivo
  concepto: string
  /** Lo marcó el sistema (GMF, comisión), no el dueño. Hoy nada lo prende. */
  automatico: boolean
  obligacion_id: number | null
  /**
   * La categoría del movimiento, ya resuelta a display por el backend. Los
   * cuatro van con `?` por la ventana de deploy (frontend y backend salen por
   * separado, y en la tablet la PWA la alarga): AUSENTE es «este servidor no
   * clasifica todavía», null es «sin clasificar» — y las dos se dibujan mudas,
   * nunca como una categoría inventada.
   */
  categoria_id?: number | null
  categoria?: string | null
  categoria_clave?: string | null
  categoria_ambito?: string | null
  /** Solo en la respuesta del POST: un aviso del backend (p.ej. una entrada de
   *  Occidente tecleada cuando las consignaciones ya entran solas). */
  advertencia?: string
  nota: string | null
}

/**
 * Una consignación PROYECTADA en el libro: entró al banco ese día, con su
 * comprobante, sin que nadie la tecleara. No es un movimiento del libro (no se
 * borra desde acá — se corrige en Consignaciones) y por eso viaja en su propia
 * lista, no adentro de `movimientos`.
 */
export interface ConsignacionLibro {
  consignacion_id: number
  tienda_id: number
  valor: number
  /** 'pendiente' = la barista la registró y el admin no la confirmó; la plata
   *  YA está en el banco igual (el comprobante es la boleta del depósito). */
  estado: 'pendiente' | 'realizada'
  barista_nombre: string | null
  imagen_url: string | null
}

/**
 * Un pago EN EFECTIVO que se muestra en su día del libro. Es INFORMATIVO:
 * esa plata salió del cajón o de la mano y nunca pasó por una cuenta, así que
 * NO está sumada en `salidas` ni mueve el saldo — sumarla rompería la
 * invariante inicial + entra − sale = final contra el extracto.
 */
export interface PagoEfectivoLibro {
  pago_id: number
  monto: number
  /** El concepto de la obligación o «Proveedor: X». null = el padre ya no
   *  existe y no se le inventa un nombre. */
  detalle: string | null
  nota: string | null
}

/** Una categoría en la serie anual de salidas: los doce meses, ene..dic. */
export interface CategoriaAnual {
  categoria_id: number | null
  clave: string | null
  /** «Sin clasificar» cuando la salida se tecleó sin categoría: un estado
   *  dicho, nunca un cero escondido. */
  nombre: string
  ambito: string | null
  meses: number[]
  total: number
}

export interface PorCategoriaAnual {
  anio: number
  /** Ordenadas por total desc por el backend: «qué me cuesta más este año». */
  categorias: CategoriaAnual[]
}

/**
 * Una fila del libro = un día. La invariante que la pantalla tiene que dejar
 * verificar a ojo, porque es la fórmula de la hoja del dueño:
 *
 *     inicial + total_entradas − total_salidas === final
 *
 * LA CADENA SE DECIDE POR DÍA, NO POR MES. `cadena` contesta la única pregunta
 * que importa para pintar un saldo: ¿se conoce el de ESTE día? Antes del ancla
 * no se sabe cuánta plata había, y el backend manda `inicial`/`final` en null
 * en vez de un número inventado (services/banco.py).
 *
 * Va tipado como UNIÓN DISCRIMINADA a propósito: `dia.cadena` es la única
 * puerta que abre el paso a los saldos, así que el compilador —y no la
 * disciplina de quien edite después— es el que impide pintar un número que el
 * backend declaró desconocido. Es exactamente el bug que se está arreglando:
 * la pantalla decidía por una bandera del MES y afirmaba saldos por fuera de
 * donde la cadena existe.
 */
interface DiaComun {
  fecha: string
  entradas: Record<string, number>   // por nombre de cuenta
  total_entradas: number
  salidas: Record<string, number>
  total_salidas: number
  /**
   * `cadena && final < 0`, calculado por el backend. No lo recalcules acá.
   * Sin cadena viene SIEMPRE en false: un saldo que no se conoce no puede
   * estar en negativo.
   */
  en_rojo: boolean
  movimientos: MovimientoBanco[]
  /**
   * Las consignaciones del día, YA sumadas en `entradas`/`total_entradas` por
   * el backend. Con `?`: un servidor de antes de la proyección no manda la
   * clave, y esa ausencia es «no sé si este día tuvo consignaciones
   * proyectadas», nunca «no tuvo».
   */
  consignaciones?: ConsignacionLibro[]
  /**
   * Los pagos EN EFECTIVO del día, para que la plata que salió del cajón o de
   * la mano se vea donde pasó. NO están en `salidas` ni en `total_salidas` y
   * no mueven el saldo. Con `?` por la ventana de deploy: ausente = «este
   * servidor no los cuenta», y no se dibuja nada.
   */
  pagos_efectivo?: PagoEfectivoLibro[]
}

/** Un día del ancla en adelante: el saldo es exacto. */
export interface DiaConSaldo extends DiaComun {
  cadena: true
  inicial: number
  final: number
}

/** Un día anterior al ancla: lo que se movió es real, el saldo no se sabe. */
export interface DiaSinSaldo extends DiaComun {
  cadena: false
  inicial: null
  final: null
}

export type DiaLibro = DiaConSaldo | DiaSinSaldo

export interface LibroMes {
  desde: string
  hasta: string
  cuentas: CuentaBanco[]
  /** El saldo del extracto que tecleó el dueño: de ahí arranca toda la cadena. */
  ancla: { saldo: number; fecha: string | null }
  /**
   * TODOS los días del rango tienen saldo.
   *
   * NO SIRVE PARA DECIDIR SI SE PINTAN LOS SALDOS. Con el ancla a mitad de mes
   * —que es el caso NORMAL, porque el editor propone hoy y el sistema pide
   * actualizar el extracto cada 7 días— esto es false y aun así del ancla en
   * adelante el saldo es exacto. Quien decide es el `cadena` de cada fila.
   */
  cadena_completa: boolean
  /** Cuántos días del rango sí tienen saldo. 0 = no hay ninguno que pintar. */
  dias_con_saldo: number
  /** Desde qué día se conoce el saldo (ISO). null = ninguno del rango. */
  primer_dia_con_saldo: string | null
  /**
   * Desde cuándo las consignaciones entran SOLAS al libro. null = el régimen
   * no se activó (todo tecleado, como siempre). AUSENTE (`undefined`) = el
   * servidor no conoce el régimen: no se ofrece activarlo, porque el botón
   * llamaría a un endpoint que no existe.
   */
  consignaciones_desde?: string | null
  /** Filas legacy sin fecha: NO están en ningún día del libro y se dicen. */
  consignaciones_sin_fecha?: { n: number; total: number }
  /** La tasa del GMF (4×1000) vigente, para SUGERIR la fila al cargar una
   *  salida. null/ausente = no se pudo leer: la sugerencia se apaga, no se
   *  inventa una tasa. */
  tasa_gmf?: number | null
  dias: DiaLibro[]
  totales: {
    entradas: number
    salidas: number
    /** null cuando el ÚLTIMO día del rango no tiene saldo: no se sabe. */
    final: number | null
    dias_en_rojo: number
    /** El cierre más bajo ENTRE LOS DÍAS CON SALDO. null si no hay ninguno. */
    dia_mas_bajo: number | null
    /** El día de ese cierre, dicho por el backend: no se busca por el monto. */
    fecha_dia_mas_bajo: string | null
  }
}

export interface MesSerie {
  mes: number                        // 1..12
  entradas: number
  salidas: number
  neto: number
  /** null = la cadena no llega hasta ese mes. NO es un cierre de cero. */
  cierre: number | null
}

export interface SerieAnual {
  anio: number
  meses: MesSerie[]
}

/** Lo que contesta `PUT /banco/ancla`, releído de la base (no lo que se mandó). */
export interface AnclaGuardada {
  saldo: number
  fecha: string | null
  dias_desde: number | null
  desactualizado: boolean
}

// ─── Fechas ───────────────────────────────────────────────────────────────────
// Todo en ISO 'YYYY-MM-DD', que es lo que devuelve el backend: comparar strings
// ISO es comparar fechas, y así ningún Date con hora corre un día por el huso.

export const MESES = ['enero', 'febrero', 'marzo', 'abril', 'mayo', 'junio',
  'julio', 'agosto', 'septiembre', 'octubre', 'noviembre', 'diciembre']

export const MESES_CORTOS = ['ene', 'feb', 'mar', 'abr', 'may', 'jun',
  'jul', 'ago', 'sep', 'oct', 'nov', 'dic']

// Fijos y no `toLocaleDateString('es-CO', { weekday: 'short' })`: el navegador de
// la tablet no siempre trae el locale y devuelve "Fri" en media pantalla.
const DIAS_CORTOS = ['dom', 'lun', 'mar', 'mié', 'jue', 'vie', 'sáb']

/** Día de la semana ('lun'…'dom') de una fecha ISO, sin pasar por UTC. */
export const diaSemana = (iso: string) => DIAS_CORTOS[new Date(iso + 'T00:00:00').getDay()]

/** Sábado o domingo: se tiñen para que el ojo agarre las semanas de un golpe. */
export const esFinde = (iso: string) => {
  const d = new Date(iso + 'T00:00:00').getDay()
  return d === 0 || d === 6
}

export const fechaLarga = (iso: string) =>
  new Date(iso + 'T00:00:00').toLocaleDateString('es-CO',
    { weekday: 'long', day: 'numeric', month: 'long' })

export const fechaCorta = (iso: string) =>
  new Date(iso + 'T00:00:00').toLocaleDateString('es-CO')

/** Días completos entre dos fechas ISO. Aritmética, no una afirmación. */
export const diasEntre = (desde: string, hasta: string) =>
  Math.round((new Date(hasta + 'T00:00:00').getTime()
    - new Date(desde + 'T00:00:00').getTime()) / 86_400_000)

// ─── Plata ────────────────────────────────────────────────────────────────────

/**
 * Plata en una celda de la grilla: solo el orden de magnitud.
 *
 * El monto exacto vive en el detalle del día, que está a un click. Una tabla de
 * 31 días con cuatro columnas de nueve dígitos cada una no se lee "de un golpe",
 * y leer el mes de un golpe es exactamente lo que esta pantalla tiene que lograr.
 */
export const compacto = (v: number) => {
  const a = Math.abs(v)
  const signo = v < 0 ? '−' : ''
  if (a >= 1_000_000) return `${signo}${(a / 1_000_000).toFixed(1).replace('.', ',')}M`
  if (a >= 1_000) return `${signo}${Math.round(a / 1_000)}k`
  return `${signo}${Math.round(a)}`
}

/**
 * Plata exacta, con el signo ADELANTE del peso: −$250.000, no $-250.000.
 *
 * El `fmt` compartido pega el signo entre el símbolo y el número porque en el
 * resto del sistema los montos no se van a negativo. Acá sí: un saldo en rojo es
 * el número que esta pantalla existe para mostrar, y «$-250.000» se lee mal justo
 * en el renglón que más importa. Mismo criterio que `compacto`, que ya usa «−».
 */
export const plata = (v: number) =>
  (v < 0 ? '−' : '') + '$' + Math.abs(Math.round(v || 0)).toLocaleString('es-CO')

// El día que cerró más bajo del mes NO se busca acá: viene en
// `totales.fecha_dia_mas_bajo`. Antes se buscaba la primera fila cuyo `final`
// fuera igual al mínimo, y esa búsqueda dejó de tener sentido cuando los días
// sin cadena pasaron a traer `final: null` — es la misma familia de error que
// este módulo viene arrastrando: identificar un dato por su FORMA (un monto que
// coincide) en vez de preguntarle al backend cuál es.

/**
 * `detalleDeError` se mudó a `src/api/errores.ts`: la necesita `useDato`, y la
 * capa de datos no puede importar de `components/`. Se re-exporta acá para que
 * los diez archivos que ya la importan desde este módulo sigan igual.
 */
export { detalleDeError } from '../../api/errores'

File: backend/app/routers/banco.py
"""El LIBRO del banco sobre HTTP: una fila por día, con la fórmula de la hoja.

Todo es admin. El saldo del banco no es información de turno: la barista no lo
necesita para operar y verlo no la ayuda a nada.

═══════════════════════════════════════════════════════════════════════════════
DIVISIÓN DE TRABAJO CON `services/banco.py`
═══════════════════════════════════════════════════════════════════════════════
La REGLA vive en el servicio; acá solo se traduce a HTTP. El servicio levanta
ValueError con un texto escrito PARA EL DUEÑO ("el monto va en positivo: el
signo lo decide si es entrada o salida"), así que ese mensaje viaja TAL CUAL
dentro del 400. Reemplazarlo por un "datos inválidos" genérico lo dejaría sin
saber qué corregir, que es justo lo que el mensaje del servicio ya resuelve.

Lo único que se valida acá y no allá son los guardrails de TECLEO que dependen
del transporte o de la columna: un `Infinity` que solo puede llegar por JSON, un
monto que no cabe en Numeric(12,2), un mes 13 que reventaría `date()`. Ninguno
de esos es una regla del negocio — son formas de que un 500 no se disfrace de
dato guardado.
"""
import math
from datetime import date
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel
from sqlalchemy.orm import Session

from app.core.deps import require_admin
from app.core.tz import hoy_col

# Piso de fecha para TODO el libro: el ancla y los movimientos. Una fecha
# anterior es siempre un tecleo (el año con dos dígitos, un dígito de más), y
# en el ANCLA pesa más que en cualquier movimiento: es la raíz de la cadena, y
# con una fecha absurdamente vieja todos los días de toda la historia quedan
# marcados con saldo confiable y el cartel que explica desde cuándo se conoce
# el saldo desaparece.
LIBRO_DESDE = date(2000, 1, 1)
from app.database import get_db
from app.models.models import Obligacion, Usuario
from app.services import banco
from app.services import costos as costos_svc

router = APIRouter(prefix="/banco", tags=["banco"])


# Cota de tecleo del año. `date(anio, mes, 1)` revienta con ValueError fuera del
# rango legal y `serie_mensual` recorre doce meses: un cero de más en el año
# convertiría un error de tipeo en un 500 sin mensaje.
ANIO_MIN, ANIO_MAX = 2000, 2100

# Lo máximo que entra en `MovimientoBanco.monto`, que es Numeric(12, 2): diez
# dígitos enteros. Más que esto lo rechaza Postgres con un error de rango —o sea
# un 500— cuando en realidad es un cero de más al teclear.
MONTO_MAX = 9_999_999_999.99


def _anio_mes(anio: Optional[int], mes: Optional[int]) -> tuple[int, int]:
    """El mes que se pide, o el de hoy en Colombia. Con 400 en vez de 500."""
    hoy = hoy_col()
    a = hoy.year if anio is None else int(anio)
    m = hoy.month if mes is None else int(mes)
    if not 1 <= m <= 12:
        raise HTTPException(400, "Mes inválido (tiene que ser 1 a 12).")
    if not ANIO_MIN <= a <= ANIO_MAX:
        raise HTTPException(400, f"Año inválido (tiene que ser {ANIO_MIN} a {ANIO_MAX}).")
    return a, m


def _anio(anio: Optional[int]) -> int:
    return _anio_mes(anio, 1)[0]


class MovimientoIn(BaseModel):
    """Un movimiento del banco, tecleado por el dueño.

    NO expone `automatico`: esa bandera marca lo que el SISTEMA sugirió solo
    (GMF, comisión del datáfono) y existe para poder distinguirlo de lo que él
    escribió a mano. Si el cliente pudiera prenderla, la distinción dejaría de
    significar nada — un movimiento que llega por este endpoint es, por
    definición, tecleado.
    """
    # Las cotas de fecha y de largo NO van acá sino en el handler, como
    # HTTPException(400). Pydantic rechaza con 422 y un `detail` que es una
    # LISTA de objetos, y el cliente solo sabe leer `detail` cuando es texto
    # (frontend/.../banco.ts): el mensaje bien escrito no llegaba nunca y el
    # dueño leía «Reintentá», que lo invita a repetir algo que va a fallar
    # siempre igual. Cambiar un 500 por un 422 mudo no es arreglarlo.
    fecha: date
    cuenta_id: int
    tipo: str                      # 'entrada' | 'salida' — lo valida el servicio
    monto: float                   # SIEMPRE positivo; el signo lo pone el tipo
    concepto: str
    # Si el movimiento paga una obligación ya cargada, se enlaza: así el
    # calendario puede tachar ese vencimiento en vez de mostrarlo pendiente.
    obligacion_id: Optional[int] = None
    # La categoría del movimiento (cualquier ámbito: café, personal, banco).
    # Opcional: un movimiento sin clasificar es válido, no un error.
    categoria_id: Optional[int] = None
    nota: Optional[str] = None


class AnclaIn(BaseModel):
    """El saldo del extracto: con cuánta plata ARRANCA ese día.

    Sin restricciones de pydantic a propósito, por el mismo motivo que
    `SaldoBancoRequest` (schemas/costos.py): un valor rechazado por el schema
    vuelve DENTRO del cuerpo del 422 y un `inf` no es serializable a JSON — la
    respuesta de error revienta antes de llegar. La validación va en el handler
    y contesta 400 con un texto que se entiende.
    """
    saldo: float
    fecha: Optional[date] = None   # None = hoy


@router.get("/libro")
def libro_del_mes(
    anio: Optional[int] = Query(None),
    mes: Optional[int] = Query(None),
    db: Session = Depends(get_db),
    admin: Usuario = Depends(require_admin),
):
    """Una fila por día del mes: arranca, entra, sale, queda.

    Van TODOS los días, también los que no tuvieron movimiento: son los que
    dejan ver que el saldo se quedó abajo cuatro días seguidos.

    LA CADENA SE DECIDE POR DÍA. Cada fila trae su `cadena`: en false el día es
    anterior al ancla, no se sabe cuánta plata había y `inicial`/`final` van en
    null. `cadena_completa` significa «TODOS los días del rango tienen saldo» y
    con el ancla a mitad de mes —el caso normal— es false aunque del ancla en
    adelante el saldo sea exacto: la pantalla NO puede decidir con esa bandera.
    Para decir la verdad parcial están `dias_con_saldo` y `primer_dia_con_saldo`.
    """
    a, m = _anio_mes(anio, mes)
    desde, hasta = banco.dias_del_mes(a, m)
    return banco.libro(db, desde, hasta)


@router.get("/por-categoria")
def por_categoria(
    anio: Optional[int] = Query(None),
    db: Session = Depends(get_db),
    admin: Usuario = Depends(require_admin),
):
    """Cuánto se está yendo en cada cosa, mes a mes: las salidas del libro
    agrupadas por categoría, los doce meses del año. Lo sin categoría viaja
    como «Sin clasificar» — dicho, no escondido."""
    a, _ = _anio_mes(anio, None)
    return banco.por_categoria_anual(db, a)


@router.get("/consignaciones-preview")
def preview_consignaciones(
    desde: date = Query(...),
    db: Session = Depends(get_db),
    admin: Usuario = Depends(require_admin),
):
    """Qué cambiaría si las consignaciones entraran solas al libro desde esa
    fecha: cuántas son y cuánta plata, y qué entradas de Occidente ya tecleadas
    en ese rango quedarían contadas DOS veces. Es la vista previa con la que se
    decide la activación — nunca un interruptor a ciegas."""
    if desde < LIBRO_DESDE:
        raise HTTPException(400, "Esa fecha es anterior al año 2000: revisá lo "
                                 "que tecleaste.")
    return banco.preview_consignaciones(db, desde)


class ConsignacionesDesdeIn(BaseModel):
    desde: date


@router.post("/consignaciones-desde")
def activar_consignaciones(
    data: ConsignacionesDesdeIn,
    db: Session = Depends(get_db),
    admin: Usuario = Depends(require_admin),
):
    """Activa el régimen: desde `desde`, cada consignación entra al libro en SU
    día, como entrada de Occidente, sin que nadie la teclee. Se fija UNA vez."""
    if data.desde < LIBRO_DESDE:
        raise HTTPException(400, "Esa fecha es anterior al año 2000: revisá lo "
                                 "que tecleaste.")
    try:
        return banco.activar_consignaciones(db, data.desde)
    except ValueError as e:
        raise HTTPException(400, str(e))


@router.get("/serie")
def serie_del_anio(
    anio: Optional[int] = Query(None),
    db: Session = Depends(get_db),
    admin: Usuario = Depends(require_admin),
):
    """Los doce meses: cuánto entró, cuánto salió y con cuánto cerró cada uno.

    `cierre` en null es la declaración honesta de que la cadena no llega hasta
    ese mes (falta el saldo del extracto): no es un cierre de cero.
    """
    return banco.serie_mensual(db, _anio(anio))


@router.get("/cuentas")
def listar_cuentas(
    solo_activas: bool = Query(True),
    db: Session = Depends(get_db),
    admin: Usuario = Depends(require_admin),
):
    """Los rieles por donde entra y sale la plata (Occidente, Bold).

    NO siembra: las cuentas las crea el arranque de la app (`main.py`). Sembrar
    de forma perezosa adentro de un GET hace que dos cargas simultáneas de la
    pantalla intenten crear la misma cuenta y una reviente con IntegrityError —
    el mismo motivo por el que los parámetros tributarios se siembran al boot.

    Misma forma que `cuentas` dentro del libro: una sola forma para la pantalla.
    """
    return [{"id": c.id, "nombre": c.nombre, "activa": bool(c.activa),
             "nota": c.nota} for c in banco.cuentas(db, solo_activas=solo_activas)]


@router.post("/movimientos")
def crear_movimiento(
    data: MovimientoIn,
    db: Session = Depends(get_db),
    admin: Usuario = Depends(require_admin),
):
    """Carga un movimiento. Devuelve el movimiento creado, NO el saldo del día.

    Y no lo devuelve a propósito: un movimiento mueve el saldo de TODOS los días
    siguientes del mes (la cadena se deriva, no se guarda). Contestar con el día
    tocado invitaría a parchear esa fila en pantalla y dejar el resto del mes
    mostrando saldos viejos. La pantalla tiene que recargar el mes.
    """
    if not math.isfinite(data.monto):
        raise HTTPException(400, "El monto tiene que ser un número válido.")
    # Solo el lado positivo: un monto negativo lo rechaza el servicio con el
    # mensaje que corresponde ("el monto va en positivo"), que es más útil que
    # mandarlo a contar ceros.
    if data.monto > MONTO_MAX:
        raise HTTPException(400, "Ese monto es demasiado grande: revisá los ceros.")
    # LA FECHA, ACOTADA ACÁ Y CON TEXTO. Sin cota, 9999-12-31 se guardaba y
    # reventaba AL RELEERLA: el libro recorre día por día y `d += timedelta(1)`
    # desborda `date.max`. Y el tope superior es HOY porque este libro es plata
    # que YA SE MOVIÓ —es lo que la propia pantalla promete— mientras que un
    # débito futuro que se sabe va en Obligaciones, que es lo único que el
    # punto de quiebre mira. Aceptándolo acá, bajaba el saldo del libro y NO
    # bajaba la proyección: dos números para la misma pregunta, con el
    # optimista en la pestaña que se mira primero.
    if data.fecha < LIBRO_DESDE:
        raise HTTPException(400, "Esa fecha es anterior al año 2000: revisá lo "
                                 "que tecleaste.")
    if data.fecha > hoy_col():
        raise HTTPException(400, "El libro es de plata que YA se movió, así que "
                                 "no acepta fechas futuras. Un débito que ya "
                                 "sabés que viene va en Obligaciones, que es lo "
                                 "que mira la proyección.")
    if len(data.concepto or "") > 160:
        raise HTTPException(400, "El concepto no puede pasar de 160 caracteres.")
    if len(data.nota or "") > 300:
        raise HTTPException(400, "La nota no puede pasar de 300 caracteres.")
    if data.obligacion_id is not None and db.query(Obligacion).filter(
            Obligacion.id == data.obligacion_id).first() is None:
        raise HTTPException(400, "Esa obligación no existe.")
    try:
        mov = banco.registrar(
            db, data.fecha, data.cuenta_id, data.tipo, data.monto, data.concepto,
            usuario_id=admin.id, obligacion_id=data.obligacion_id, nota=data.nota,
            categoria_id=data.categoria_id)
    except ValueError as e:
        # El texto del servicio ya está escrito para que lo lea el dueño.
        raise HTTPException(400, str(e))

    # Se relee por el libro de ESE día en vez de armar el dict acá: así el
    # movimiento que devuelve el POST tiene exactamente la misma forma que los
    # que vienen adentro de `dias[].movimientos`, y la pantalla no necesita
    # aprender dos formas del mismo objeto.
    fila = banco.libro(db, mov.fecha, mov.fecha)["dias"][0]
    out = next(m for m in fila["movimientos"] if m["id"] == mov.id)
    # Bajo el régimen de consignaciones-al-libro, una entrada de Occidente
    # tecleada puede ser la MISMA plata que una consignación ya proyectada.
    # No se bloquea (una transferencia recibida es legítima) pero se dice, con
    # la clave aditiva que el bundle viejo ignora.
    corte = banco.consignaciones_desde(db)
    if (corte is not None and data.tipo == "entrada" and data.fecha >= corte
            and out.get("cuenta") == "Occidente"):
        out["advertencia"] = (
            "Ojo: desde el " + corte.isoformat() + " las consignaciones entran "
            "solas al libro. Si esta entrada es una consignación, ya está "
            "contada y quedaría dos veces — borrala si es el caso.")
    return out


@router.delete("/movimientos/{movimiento_id}")
def borrar_movimiento(
    movimiento_id: int,
    db: Session = Depends(get_db),
    admin: Usuario = Depends(require_admin),
):
    """Se borra de verdad y el saldo se reacomoda solo (está derivado).

    No hay "anulado" en este libro: un movimiento anulado que siguiera sumando
    sería un saldo mentiroso, y uno que no sumara es idéntico a no existir.
    """
    if not banco.borrar(db, movimiento_id):
        raise HTTPException(404, "Ese movimiento no existe.")
    return {"ok": True}


@router.put("/ancla")
def declarar_ancla(
    data: AnclaIn,
    db: Session = Depends(get_db),
    admin: Usuario = Depends(require_admin),
):
    """El saldo del extracto, que es de dónde arranca toda la cadena.

    Es un dato del dueño: el sistema registra consignaciones, nunca un saldo
    bancario, así que no puede derivarlo. Sin él NINGÚN día del libro tiene
    saldo (`dias_con_saldo` en 0), y con él lo tienen los días de esa fecha en
    adelante — los anteriores siguen en null, que es lo honesto.

    ESCRIBE LAS MISMAS DOS CLAVES que `POST /costos/saldo-banco`
    (`saldo_banco` y `saldo_banco_fecha` en `configuracion`), y lo hace llamando
    al MISMO servicio: dos escritores con reglas distintas sobre la misma fila
    darían dos verdades según por qué pantalla se entró. Por eso las
    validaciones de abajo son las de aquel handler, no unas nuevas.
    """
    if not math.isfinite(data.saldo):
        raise HTTPException(400, "El saldo del banco tiene que ser un número válido.")
    if data.saldo < 0:
        raise HTTPException(400, "El saldo del banco no puede ser negativo.")
    if data.saldo > costos_svc.SALDO_BANCO_MAX:
        raise HTTPException(400, "El saldo del banco es demasiado grande.")
    fecha = data.fecha or hoy_col()
    if fecha < LIBRO_DESDE:
        raise HTTPException(400, "La fecha del extracto es anterior al año 2000: "
                                     "revisá lo que tecleaste.")
    if fecha > hoy_col():
        raise HTTPException(400, "La fecha del saldo no puede ser futura.")

    costos_svc.guardar_saldo_banco(db, data.saldo, fecha, admin.id)

    # Se relee del ancla (no se devuelve lo que llegó): lo que la pantalla
    # muestre tiene que ser lo que quedó guardado, no lo que se pidió guardar.
    saldo, guardada = banco.ancla(db)
    return {
        "saldo": saldo,
        "fecha": guardada.isoformat() if guardada else None,
        # Cuántos días hace que se miró ese extracto. Es el dato que deja decir
        # "saldo del viernes" en vez de hacerlo pasar por el de hoy.
        "dias_desde": (hoy_col() - guardada).days if guardada else None,
        "desactualizado": (guardada is None
                           or (hoy_col() - guardada).days
                           > costos_svc.DIAS_SALDO_BANCO_VIGENTE),
    }
